src/bam/record.rs

```diff
@@ -1124,45 +1121,87 @@ impl TryFrom<&[u8]> for CigarString {
-                    "S" => Cigar::SoftClip(n),
-                    "P" => Cigar::Pad(n),
-                    "=" => Cigar::Equal(n),
-                    "X" => Cigar::Diff(n),
-                    op => {
-                        return Err(Error::BamParseCigar {
-                            msg: format!("operation {} not expected", op),
-                        });
-                    }
-                });
-            } else {
+        let text_len = text.len();
+        while i < text_len {
+            let mut j = i;
+            while j < text_len && bytes[j].is_ascii_digit() {
+                j += 1;
+            }
+            // check that length is provided
+            if i == j {
                 return Err(Error::BamParseCigar {
-                    msg: "expected cigar operation [0-9]+[MIDNSHP=X]".to_owned(),
+                    msg: "Expected length before cigar operation [0-9]+[MIDNSHP=X]".to_owned(),
                 });
             }
+            // get the length of the operation
+            let n = u32::from_str(&text[i..j]).map_err(|_| Error::BamParseCigar {
+                msg: format!("Unable to parse &str '{}' to u32.", &text[i..j]),
+            })?;
+            // get the operation
+            let op = &text[j..j + 1];
+            inner.push(match op {
+                "M" => Cigar::Match(n),
+                "I" => Cigar::Ins(n),
+                "D" => Cigar::Del(n),
+                "N" => Cigar::RefSkip(n),
+                "H" => {
+                    if i == 0 || j + 1 == text_len {
+                        Cigar::HardClip(n)
+                    } else {
+                        return Err(Error::BamParseCigar {
+                            msg: "Hard clipping ('H') is only valid at the start or end of a cigar."
+                                .to_owned(),
+                        });
+                    }
+                }
+                "S" => {
+                    if i == 0
+                        || j+1 == text_len
+                        || &text[i-1..i] == "H"
+                        || text[j+1..].chars().all(|c| c.is_ascii_digit() || c == 'H') {
+                        Cigar::SoftClip(n)
+                    } else {
+                        return Err(Error::BamParseCigar {
+                        msg: "Soft clips ('S') can only have hard clips ('H') between them and the end of the CIGAR string."
+                            .to_owned(),
+                        });
+                    }
+                },
+                "P" => Cigar::Pad(n),
+                "=" => Cigar::Equal(n),
+                "X" => Cigar::Diff(n),
+                op => {
+                    return Err(Error::BamParseCigar {
+                        msg: format!("Expected cigar operation [MIDNSHP=X] but got [{}]", op),
+                    })
+                }
+            });
+            i = j + 1;
         }
-
         Ok(CigarString(inner))
     }
 }
```

Comment:
```suggestion
}

impl TryFrom<&str> for CigarString {
    type Error = Error;

    /// Create a CigarString from given &str.
    /// # Example
    /// ```
    /// use rust_htslib::bam::record::*;
    /// use rust_htslib::bam::record::CigarString;
    /// use rust_htslib::bam::record::Cigar::*;
    /// use std::convert::TryFrom;
    ///
    /// let cigar_str = "2H10M5X3=2H";
    /// let cigar = CigarString::try_from(cigar_str)
    ///     .expect("Unable to parse cigar string.");
    /// let expected_cigar = CigarString(vec![
    ///     HardClip(2),
    ///     Match(10),
    ///     Diff(5),
    ///     Equal(3),
    ///     HardClip(2),
    /// ]);
    /// assert_eq!(cigar, expected_cigar);
    /// ```
    fn try_from(text: &str) -> Result<Self> {
        let bytes = text.as_bytes();
        if ( text.chars().count() == bytes.len() ) {
            return Err(Error::BamParseCigar {
                msg: "CIGAR string contained non-ASCII characters, which are not valid. Valid are [0-9MIDNSHP=X].".to_owned(),
            });
        }
        CigarString::try_from(bytes)
    }
}